File: orchestrator/graph/schemas/template_registry.py
"""
Template Registry

Central source of truth for document templates across all formats.
Python equivalent of frontend/src/lib/orchestrator/schemas/templateRegistry.ts
"""

from typing import Optional, List, Dict, Any

class Template:
    """Template definition"""
    def __init__(
        self,
        id: str,
        name: str,
        description: str,
        keywords: Optional[List[str]] = None,
        complexity: str = 'simple',  # simple | moderate | complex
        recommended_for: Optional[str] = None
    ):
        self.id = id
        self.name = name
        self.description = description
        self.keywords = keywords or []
        self.complexity = complexity
        self.recommended_for = recommended_for
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'keywords': self.keywords,
            'complexity': self.complexity,
            'recommendedFor': self.recommended_for
        }


# Template Registry - Single Source of Truth
TEMPLATE_REGISTRY: Dict[str, List[Template]] = {
    'novel': [
        Template('three-act', 'Three-Act Structure', 'Classic beginning, middle, and end',
                 ['three act', 'classic', 'traditional', 'beginning middle end'],
                 'simple', 'Traditional storytelling with clear setup, confrontation, and resolution'),
        Template('heros-journey', "Hero's Journey", 'Archetypal adventure narrative',
                 ['hero', 'journey', 'monomyth', 'campbell', 'adventure', 'quest'],
                 'moderate', 'Epic adventures, fantasy, or transformative character arcs'),
        Template('freytag', "Freytag's Pyramid", 'Rising action, climax, falling action',
                 ['freytag', 'pyramid', 'rising action', 'climax', 'falling action'],
                 'moderate', 'Dramatic stories with clear tension build-up and resolution'),
        Template('save-the-cat', 'Save The Cat', 'Modern screenplay structure adapted for novels',
                 ['save the cat', 'snyder', 'beat sheet', 'screenplay'],
                 'complex', 'Plot-driven novels with commercial appeal'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
    'short-story': [
        Template('classic', 'Classic Short Story', 'Single plot, few characters, brief timespan',
                 ['classic', 'traditional', 'single plot'],
                 'simple', 'Traditional short fiction with focused narrative'),
        Template('flash-fiction', 'Flash Fiction', 'Ultra-short 500-1000 words',
                 ['flash', 'micro', 'ultra short', 'brief'],
                 'simple', 'Extremely concise stories with punchy impact'),
        Template('twist-ending', 'Twist Ending', 'Surprise revelation structure',
                 ['twist', 'surprise', 'revelation', 'unexpected'],
                 'moderate', 'Stories with shocking or unexpected conclusions'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
    'screenplay': [
        Template('feature', 'Feature Film', '90-120 pages, three acts',
                 ['feature', 'film', 'movie', 'theatrical', '90 minutes', 'three act'],
                 'complex', 'Full-length theatrical films'),
        Template('tv-pilot', 'TV Pilot', '30 or 60-minute episode',
                 ['tv', 'television', 'pilot', 'series', 'episode'],
                 'complex', 'Television series premiere episodes'),
        Template('short-film', 'Short Film', '5-30 pages',
                 ['short', 'short film', 'brief', 'festival'],
                 'moderate', 'Festival submissions or proof-of-concept'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
    'podcast': [
        Template('interview', 'Interview Format', 'Host interviews guests',
                 ['interview', 'guest', 'host', 'q&a', 'conversation'],
                 'simple', 'Guest-focused conversational podcasts'),
        Template('co-hosted', 'Co-Hosted', 'Multiple hosts in conversation',
                 ['co-hosted', 'multiple hosts', 'panel', 'discussion'],
                 'moderate', 'Dynamic multi-host discussions'),
        Template('storytelling', 'Storytelling', 'Narrative-driven episodes',
                 ['storytelling', 'narrative', 'story', 'documentary'],
                 'complex', 'Narrative or documentary-style podcasts'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
    'report': [
        Template('business', 'Business Report', 'Executive summary, findings, recommendations',
                 ['business', 'executive', 'corporate', 'findings'],
                 'moderate', 'Corporate analysis and strategic recommendations'),
        Template('research', 'Research Report', 'Literature review, methodology, results',
                 ['research', 'academic', 'methodology', 'literature review'],
                 'complex', 'Academic or scientific research documentation'),
        Template('technical', 'Technical Report', 'Specifications, analysis, documentation',
                 ['technical', 'specs', 'documentation', 'analysis'],
                 'complex', 'Technical specifications and system documentation'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
    'article': [
        Template('how-to', 'How-To Guide', 'Step-by-step instructional',
                 ['how to', 'guide', 'tutorial', 'instructional', 'steps'],
                 'simple', 'Instructional content with clear steps'),
        Template('listicle', 'Listicle', 'Numbered or bulleted list format',
                 ['list', 'listicle', 'numbered', 'top 10', 'bullets'],
                 'simple', 'Scannable content with multiple points'),
        Template('opinion', 'Opinion Piece', 'Editorial or commentary',
                 ['opinion', 'editorial', 'commentary', 'perspective'],
                 'moderate', 'Persuasive or thought-provoking content'),
        Template('feature', 'Feature Article', 'In-depth exploration of topic',
                 ['feature', 'in-depth', 'long-form', 'deep dive'],
                 'complex', 'Comprehensive topic exploration'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
    'essay': [
        Template('argumentative', 'Argumentative', 'Claim, evidence, counterarguments',
                 ['argumentative', 'persuasive', 'debate', 'claim', 'evidence'],
                 'moderate', 'Persuasive essays with clear thesis'),
        Template('narrative', 'Narrative Essay', 'Personal story with reflection',
                 ['narrative', 'personal', 'story', 'reflection', 'memoir'],
                 'simple', 'Personal storytelling with insight'),
        Template('compare-contrast', 'Compare & Contrast', 'Analyze similarities and differences',
                 ['compare', 'contrast', 'comparison', 'similarities', 'differences'],
                 'moderate', 'Analytical comparison of subjects'),
        Template('blank', 'Blank Canvas', 'Start from scratch',
                 ['blank', 'custom', 'freeform', 'scratch'],
                 'simple', 'Complete creative freedom'),
    ],
}


def get_templates_for_format(format_type: str) -> List[Template]:
    """Get templates for a specific format"""
    return TEMPLATE_REGISTRY.get(format_type, [])


def get_template_by_id(format_type: str, template_id: str) -> Optional[Template]:
    """Get a specific template by ID"""
    templates = get_templates_for_format(format_type)
    for template in templates:
        if template.id == template_id:
            return template
    return None


def find_template_by_keywords(format_type: str, query: str) -> Optional[Template]:
    """Find template by keyword matching (for intent analysis)"""
    templates = get_templates_for_format(format_type)
    lower_query = query.lower()
    
    # Try exact name match first
    for template in templates:
        if template.name.lower() == lower_query:
            return template
    
    # Try keyword matching
    for template in templates:
        for keyword in template.keywords:
            if keyword in lower_query or lower_query in keyword:
                return template
    
    return None


def get_format_label(format_type: str) -> str:
    """Get human-readable label for format"""
    labels = {
        'novel': 'Novel',
        'short-story': 'Short Story',
        'screenplay': 'Screenplay',
        'podcast': 'Podcast',
        'report': 'Report',
        'article': 'Article',
        'essay': 'Essay',
    }
    return labels.get(format_type, format_type.replace('-', ' ').title())


def get_available_formats() -> List[str]:
    """Get all formats that have templates"""
    return list(TEMPLATE_REGISTRY.keys())

File: frontend/src/lib/orchestrator/schemas/templateRegistry.ts
/**
 * Template Registry
 * 
 * Central source of truth for document templates across all formats.
 * Used by:
 * - LLM Intent Analyzer (for intelligent template suggestions)
 * - CreateStructureAction (for structure generation)
 * - OrchestratorPanel (for UI display)
 */

export interface Template {
  id: string
  name: string
  description: string
  keywords?: string[] // For LLM matching (e.g., ["interview", "guest", "host"])
  complexity?: 'simple' | 'moderate' | 'complex' // For user guidance
  recommendedFor?: string // When to use this template
}

export interface TemplateCategory {
  format: string
  label: string
  templates: Template[]
}

export type TemplateRegistry = Record<string, Template[]>

/**
 * Template Registry - Single Source of Truth
 */
export const TEMPLATE_REGISTRY: TemplateRegistry = {
  'novel': [
    {
      id: 'three-act',
      name: 'Three-Act Structure',
      description: 'Classic beginning, middle, and end',
      keywords: ['three act', 'classic', 'traditional', 'beginning middle end'],
      complexity: 'simple',
      recommendedFor: 'Traditional storytelling with clear setup, confrontation, and resolution'
    },
    {
      id: 'heros-journey',
      name: "Hero's Journey",
      description: 'Archetypal adventure narrative',
      keywords: ['hero', 'journey', 'monomyth', 'campbell', 'adventure', 'quest'],
      complexity: 'moderate',
      recommendedFor: 'Epic adventures, fantasy, or transformative character arcs'
    },
    {
      id: 'freytag',
      name: "Freytag's Pyramid",
      description: 'Rising action, climax, falling action',
      keywords: ['freytag', 'pyramid', 'rising action', 'climax', 'falling action'],
      complexity: 'moderate',
      recommendedFor: 'Dramatic stories with clear tension build-up and resolution'
    },
    {
      id: 'save-the-cat',
      name: 'Save The Cat',
      description: 'Modern screenplay structure adapted for novels',
      keywords: ['save the cat', 'snyder', 'beat sheet', 'screenplay'],
      complexity: 'complex',
      recommendedFor: 'Plot-driven novels with commercial appeal'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'short-story': [
    {
      id: 'classic',
      name: 'Classic Short Story',
      description: 'Single plot, few characters, brief timespan',
      keywords: ['classic', 'traditional', 'single plot'],
      complexity: 'simple',
      recommendedFor: 'Traditional short fiction with focused narrative'
    },
    {
      id: 'flash-fiction',
      name: 'Flash Fiction',
      description: 'Ultra-short 500-1000 words',
      keywords: ['flash', 'micro', 'ultra short', 'brief'],
      complexity: 'simple',
      recommendedFor: 'Extremely concise stories with punchy impact'
    },
    {
      id: 'twist-ending',
      name: 'Twist Ending',
      description: 'Surprise revelation structure',
      keywords: ['twist', 'surprise', 'revelation', 'unexpected'],
      complexity: 'moderate',
      recommendedFor: 'Stories with shocking or unexpected conclusions'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'report': [
    {
      id: 'business',
      name: 'Business Report',
      description: 'Executive summary, findings, recommendations',
      keywords: ['business', 'executive', 'corporate', 'findings'],
      complexity: 'moderate',
      recommendedFor: 'Corporate analysis and strategic recommendations'
    },
    {
      id: 'research',
      name: 'Research Report',
      description: 'Literature review, methodology, results',
      keywords: ['research', 'academic', 'methodology', 'literature review'],
      complexity: 'complex',
      recommendedFor: 'Academic or scientific research documentation'
    },
    {
      id: 'technical',
      name: 'Technical Report',
      description: 'Specifications, analysis, documentation',
      keywords: ['technical', 'specs', 'documentation', 'analysis'],
      complexity: 'complex',
      recommendedFor: 'Technical specifications and system documentation'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'report_script_coverage': [
    {
      id: 'standard',
      name: 'Standard Coverage',
      description: 'Industry-standard screenplay analysis',
      keywords: ['standard', 'industry', 'screenplay', 'coverage'],
      complexity: 'moderate',
      recommendedFor: 'Professional screenplay evaluation for production companies'
    },
    {
      id: 'detailed',
      name: 'Detailed Coverage',
      description: 'In-depth analysis with recommendations',
      keywords: ['detailed', 'in-depth', 'comprehensive', 'analysis'],
      complexity: 'complex',
      recommendedFor: 'Thorough screenplay analysis with development notes'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'report_business': [
    {
      id: 'executive',
      name: 'Executive Report',
      description: 'High-level strategic analysis',
      keywords: ['executive', 'strategic', 'high-level', 'leadership'],
      complexity: 'moderate',
      recommendedFor: 'C-suite and board-level strategic reports'
    },
    {
      id: 'analytical',
      name: 'Analytical Report',
      description: 'Data-driven insights and recommendations',
      keywords: ['analytical', 'data', 'insights', 'metrics'],
      complexity: 'complex',
      recommendedFor: 'Data-driven business intelligence reports'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'report_content_analysis': [
    {
      id: 'thematic',
      name: 'Thematic Analysis',
      description: 'Focus on themes and patterns',
      keywords: ['thematic', 'themes', 'patterns', 'motifs'],
      complexity: 'moderate',
      recommendedFor: 'Analyzing recurring themes and narrative patterns'
    },
    {
      id: 'structural',
      name: 'Structural Analysis',
      description: 'Analyze content structure and flow',
      keywords: ['structural', 'structure', 'flow', 'organization'],
      complexity: 'moderate',
      recommendedFor: 'Evaluating narrative structure and pacing'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'article': [
    {
      id: 'how-to',
      name: 'How-To Guide',
      description: 'Step-by-step instructional',
      keywords: ['how to', 'guide', 'tutorial', 'instructional', 'steps'],
      complexity: 'simple',
      recommendedFor: 'Instructional content with clear steps'
    },
    {
      id: 'listicle',
      name: 'Listicle',
      description: 'Numbered or bulleted list format',
      keywords: ['list', 'listicle', 'numbered', 'top 10', 'bullets'],
      complexity: 'simple',
      recommendedFor: 'Scannable content with multiple points'
    },
    {
      id: 'opinion',
      name: 'Opinion Piece',
      description: 'Editorial or commentary',
      keywords: ['opinion', 'editorial', 'commentary', 'perspective'],
      complexity: 'moderate',
      recommendedFor: 'Persuasive or thought-provoking content'
    },
    {
      id: 'feature',
      name: 'Feature Article',
      description: 'In-depth exploration of topic',
      keywords: ['feature', 'in-depth', 'long-form', 'deep dive'],
      complexity: 'complex',
      recommendedFor: 'Comprehensive topic exploration'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'screenplay': [
    {
      id: 'feature',
      name: 'Feature Film',
      description: '90-120 pages, three acts',
      keywords: ['feature', 'film', 'movie', 'theatrical', '90 minutes', 'three act'],
      complexity: 'complex',
      recommendedFor: 'Full-length theatrical films'
    },
    {
      id: 'tv-pilot',
      name: 'TV Pilot',
      description: '30 or 60-minute episode',
      keywords: ['tv', 'television', 'pilot', 'series', 'episode'],
      complexity: 'complex',
      recommendedFor: 'Television series premiere episodes'
    },
    {
      id: 'short-film',
      name: 'Short Film',
      description: '5-30 pages',
      keywords: ['short', 'short film', 'brief', 'festival'],
      complexity: 'moderate',
      recommendedFor: 'Festival submissions or proof-of-concept'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'essay': [
    {
      id: 'argumentative',
      name: 'Argumentative',
      description: 'Claim, evidence, counterarguments',
      keywords: ['argumentative', 'persuasive', 'debate', 'claim', 'evidence'],
      complexity: 'moderate',
      recommendedFor: 'Persuasive essays with clear thesis'
    },
    {
      id: 'narrative',
      name: 'Narrative Essay',
      description: 'Personal story with reflection',
      keywords: ['narrative', 'personal', 'story', 'reflection', 'memoir'],
      complexity: 'simple',
      recommendedFor: 'Personal storytelling with insight'
    },
    {
      id: 'compare-contrast',
      name: 'Compare & Contrast',
      description: 'Analyze similarities and differences',
      keywords: ['compare', 'contrast', 'comparison', 'similarities', 'differences'],
      complexity: 'moderate',
      recommendedFor: 'Analytical comparison of subjects'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ],
  'podcast': [
    {
      id: 'interview',
      name: 'Interview Format',
      description: 'Host interviews guests',
      keywords: ['interview', 'guest', 'host', 'q&a', 'conversation'],
      complexity: 'simple',
      recommendedFor: 'Guest-focused conversational podcasts'
    },
    {
      id: 'co-hosted',
      name: 'Co-Hosted',
      description: 'Multiple hosts in conversation',
      keywords: ['co-hosted', 'multiple hosts', 'panel', 'discussion'],
      complexity: 'moderate',
      recommendedFor: 'Dynamic multi-host discussions'
    },
    {
      id: 'storytelling',
      name: 'Storytelling',
      description: 'Narrative-driven episodes',
      keywords: ['storytelling', 'narrative', 'story', 'documentary'],
      complexity: 'complex',
      recommendedFor: 'Narrative or documentary-style podcasts'
    },
    {
      id: 'blank',
      name: 'Blank Canvas',
      description: 'Start from scratch',
      keywords: ['blank', 'custom', 'freeform', 'scratch'],
      complexity: 'simple',
      recommendedFor: 'When you want complete creative freedom'
    }
  ]
}

/**
 * Get templates for a specific format
 */
export function getTemplatesForFormat(format: string): Template[] {
  return TEMPLATE_REGISTRY[format] || []
}

/**
 * Get a specific template by ID
 */
export function getTemplateById(format: string, templateId: string): Template | null {
  const templates = getTemplatesForFormat(format)
  return templates.find(t => t.id === templateId) || null
}

/**
 * Find template by keyword matching (for LLM intent analysis)
 */
export function findTemplateByKeywords(format: string, query: string): Template | null {
  const templates = getTemplatesForFormat(format)
  const lowerQuery = query.toLowerCase()
  
  // Try exact name match first
  const exactMatch = templates.find(t => t.name.toLowerCase() === lowerQuery)
  if (exactMatch) return exactMatch
  
  // Try keyword matching
  for (const template of templates) {
    if (template.keywords) {
      for (const keyword of template.keywords) {
        if (lowerQuery.includes(keyword) || keyword.includes(lowerQuery)) {
          return template
        }
      }
    }
  }
  
  return null
}

/**
 * Build template descriptions for LLM prompts
 * Similar to buildFormatDescriptionsForLLM() in documentHierarchy.ts
 */
export function buildTemplateDescriptionsForLLM(): string {
  let descriptions = 'AVAILABLE TEMPLATES BY FORMAT:\n\n'
  
  for (const [format, templates] of Object.entries(TEMPLATE_REGISTRY)) {
    const formatLabel = format.replace(/_/g, '-')
    descriptions += `${formatLabel}:\n`
    
    templates.forEach(template => {
      descriptions += `  - ${template.name}: ${template.description}`
      if (template.recommendedFor) {
        descriptions += ` (Best for: ${template.recommendedFor})`
      }
      descriptions += '\n'
    })
    descriptions += '\n'
  }
  
  descriptions += 'TEMPLATE MATCHING:\n'
  descriptions += '- If user mentions specific keywords (e.g., "interview", "hero\'s journey", "three act"), suggest the matching template\n'
  descriptions += '- If user is vague (e.g., "create a podcast"), ask which template they prefer\n'
  descriptions += '- Always include template suggestions in extractedEntities.suggestedTemplate\n'
  
  return descriptions
}

/**
 * Get all formats that have templates
 */
export function getAvailableFormats(): string[] {
  return Object.keys(TEMPLATE_REGISTRY)
}

/**
 * Get template count for a format
 */
export function getTemplateCount(format: string): number {
  return getTemplatesForFormat(format).length
}

